### ISSUE TITLE
Combined eBGP/iBGP topology (multiple AS per node)

### ISSUE THREAD

**Issue by jbemmel** [ipspace/netsim-tools#163]
I tried to create a fabric using both iBGP (for an EVPN overlay) and eBGP (as IGP):
```
bgp:
  as: 65000 # Primary AS (iBGP): overlay (for auto-RD/RT to work correctly)
  as_list:
    65000: # EVPN overlay
      members: [ leaf1a,leaf1b,leaf2a,leaf2b,spine1,spine2,h1,h2 ]
      rr: [ spine1,spine2 ]

    # eBGP underlay
    65001:
      members: [ leaf1a ]
    65002:
      members: [ leaf1b ]
    65003:
      members: [ leaf2a ]
    65004:
      members: [ leaf2b ]
    65005:
      members: [ spine1,spine2 ]
    65100:
      members: [ isp ]
```

Currently, this does not work as there is no way to mark an AS as 'primary', or differentiate between an AS intended for iBGP peering versus eBGP. Moreover, an attempt is made to peer between the route reflectors, which I wouldn't expect / didn't want to happen

How would you like to tackle this?

**Comment by ipspace** [comment]
Let's start with me being strongly opposed to the "run IBGP over EBGP" idea. Search my EVPN-related blog posts or https://www.ipspace.net/kb/tag/BGP-DC.html for details.

Forgetting all that, obviously I have no idea how SR OS or SR Linux handles this, but every other platform needs some crazy **local-as** \*\*\*\* to make it work.  I have no idea how to handle this with a sane BGP data model, maybe you could hide the IBGP details into "evpn" top-level attribute or something.

**Comment by jbemmel** [comment]
I'm well aware of your views - I figured I'd threaten SRv6, then offer to take that off the table in exchange for iBGP-over-eBGP...

I'm ok with ```evpn.as``` (and implied iBGP)

**Comment by jbemmel** [comment]
In the above as_list example with spine1/spine2 being members of both AS 65000 and AS 65005, with the current bgp module code the 65000 details are lost (overwritten by 65005).

We can have the evpn module be opiniated and correct the bgp module's limitations, allowing for multiple AS per node and adding local-as where needed (overwriting part of the config put in place by bgp)

Question on RR-RR peering, in most designs I'm familiar with that is typically avoided?

**Comment by ipspace** [comment]
> We can have the evpn module be opiniated and correct the bgp module's limitations, allowing for multiple AS per node and adding local-as where needed (overwriting part of the config put in place by bgp)

I don't think I'll ever accept a PR for an EVPN module that implements IBGP over EBGP. If someone desperately wishes to create that monstrosity they're most welcome to create a plugin to implement that. I will not waste everyone's time by implementing a heptagonal star-shaped peg force-hammered into a triangle hole just because some vendors felt the need to be hip and pushed draft-lapukhov way beyond what it was meant to do while trying desperately trying to hide the bugs in their implementation. See https://blog.ipspace.net/2020/02/the-evpnbgp-saga-continues.html and https://blog.ipspace.net/2019/04/dont-sugarcoat-challenges-you-have.html for details.

> Question on RR-RR peering, in most designs I'm familiar with that is typically avoided?

BGP was originally used in networks that were not highly symmetrical leaf-and-spine fabrics with no external connectivity on spines ;)

**Comment by jbemmel** [comment]
While EVPN triggered the issue, it does not require IBGP-over-EBGP - just BGP in general. I'll separate EVPN from this combined eBGP/iBGP discussion, and try to build the module in a way that doesn't rub you in the wrong way...

That being said, while working on EVPN I implemented some changes that would allow the bgp module to support ibgp-over-ebgp topologies; it would fix the loss of information mentioned above. It's a generalization of the internal data model - is that something you would be willing to consider?

**Comment by ipspace** [comment]
> While EVPN triggered the issue, it does not require IBGP-over-EBGP - just BGP in general.

Correct, and there are two sane ways of running it. EBGP everywhere, or IBGP over IGP. If you're considering EVPN module, I'd suggest to go for IBGP-over-IGP as it's easier to implement.

> I'll separate EVPN from this combined eBGP/iBGP discussion, and try to build the module in a way that doesn't rub you in the wrong way...

I know we don't agree, but some things are plain wrong, no matter how much lipstick vendor marketing slaps over them.

> That being said, while working on EVPN I implemented some changes that would allow the bgp module to support ibgp-over-ebgp topologies; it would fix the loss of information mentioned above. It's a generalization of the internal data model - is that something you would be willing to consider?

There is no loss of information, and there is nothing to fix. BGP was designed (and in most cases implemented) to have a single autonomous system per box. Vendor marketing engineers trying to make their software work in situations it was never designed to encounter usually come up with crazy ideas like "local AS everywhere", see https://blog.ipspace.net/2018/05/dissecting-ibgpebgp-junos-configuration.html for an example.

You're most welcome to implement a plugin that will support BGP **local-as** or whatever else it's called in any particular implementation, or build that into ibgp-ebgp-evpn plugin... and I will stop wasting everyone's time with this discussion.

**Comment by jbemmel** [comment]
> I know we don't agree, but some things are plain wrong, no matter how much lipstick vendor marketing slaps over them.

It's not that we "don't agree", I can see you have your reasons and I'm even willing to accept your position that iBGP-over-eBGP is best avoided - as I have no practical experience to the contrary.

My problem is that iBGP-over-eBGP deployments do exist, and some customers are used to doing things that way.
TL;DR I guess I'll build it as plugins
